README.md
Featureless job queue
======================

Very simple job queue focusing on high performance / high throughput.

Basically, just a distributed FIFO.

## Why use this library?
There are very decent job queues for more complex use cases (TTL, job retries, priority, progress indicator). If you're interested in this kind of features, have a look at [Kue](https://github.com/Automattic/kue).

However, those features come with a tradeoff -- tons of calls to Redis' HSET, potential stuck jobs spoiling concurrency, and `O(log(n))` operations to pop something in a priority queue.

Moreover, those libraries need to maintain a bigger pool of connection to Redis, which can be costly on Redis providers billing by the number of open connections.

They also store atomic data in multiple keys, which lead to an increased memory consumption and number of calls to Redis.

By only focusing on one use case, this library is much faster but also less flexible. See [this page](https://github.com/Neamar/featureless-job-queue/issues/1) for a performance comparison between kue and this library.

## When should I use this library?

* You need to run many small jobs at once
* Your tasks are not CPU bound and you can run more than your number of CPU cores
* You need concurrency across many servers (if you don't, just use [`async.queue`](https://caolan.github.io/async/docs.html#queue))
* You don't need priority -- just a plain FIFO
* You can afford to lose a couple of tasks on critical app failure (SIGKILL or power failure). SIGTERM is fine.

## How to use this library
> Too lazy to read doc? Have a look at the [examples folder](https://github.com/Neamar/featureless-job-queue/tree/master/examples)!

### Constructor
```js
var FJQ = require('featureless-job-queue')
var fjq = new FJQ({redisUrl: redisUrl})
```

The constructor accepts two important options:

* `redisUrl`, the URL to use for Redis. If unspecified, this will default to localhost on the default port.
* `redisKey`, the redis key to use to store the jobs. Default is `fjq:jobs`, change it if you need to run multiple job queues on the same Redis instance.

Other options are documented lower in this README, where it makes sense to introduce them.

### Process jobs
```js
var FJQ = require('featureless-job-queue')
var fjq = new FJQ({redisUrl: redisUrl})

// Number of workers to run at the same time, will default to 1
var concurrency = 40
var workerFunction = function(job, cb) {
    // your function, call cb(err) once done.    
}

fjq.process(workerFunction, concurrency)
```

Concurrency will always be respected, but note that some jobs might be unqueued before they're sent to a worker to ensure optimal throughput. This behavior can be tweaked by specifying the key `overfillRatio` in the constructor options. This value defaults to 1.1 (e.g., for concurrency of 40, 40 workers will run in parallel, and 4 tasks will be pre-buffered to be sent to workers).

This function will return an [`async.queue`](https://caolan.github.io/async/docs.html#queue). You can listen to events on it, and FJQ will append jobs to the queue automatically. Updating the queue concurrency in real time is not supported.

### Queue jobs
```js
var FJQ = require('featureless-job-queue')
var fjq = new FJQ({redisUrl: redisUrl})

// job can be any valid JS structure.
// Try to keep it as small as possible since it will transit across the network and be fully stored in Redis
// It will be serialized using JSON.serialize, so you can't use any fancy items in your job (e.g. functions)
var job = {
    type: "something",
    foo: "bar"
}

fjq.create(job, function(err, count) {
    if(err) return console.warn(err)
    // Job successfully created
    // `count` is the number of items currently in the queue (after saving this job)
})

// OR
var jobs = [{job1: true}, {job2: true}, ...]
fjq.create(job, function(err, count) {
    // Please note: if you get an error, some jobs may have been saved and others not :(
    if(err) return console.warn(err)
    // Jobs successfully created
})
```

If you need to save multiple jobs at once, use the array version of the function to minimize the number of calls sent to Redis. There is no limit to the number of jobs that can be saved at once.

### Stop processing
```js
fjq.shutdown(function() {
    console.log("All workers drained, Redis connections closed!")
})
```

The callback will only be called once all of this is true:

* all workers have finished their current task (including tasks pre-buffered with `overfillRatio`). *Note that this means some workers might start new jobs even after you've called `.shutdown`!*
* all opened Redis connections have been closed

You'll probably want to hook this function to `process.on('SIGTERM')` ;)

### Other
* use `fjq.length(function(err, count))` to get the number of jobs currently queued in Redis (doesn't include tasks currently worked on and pre-buffered tasks)
* use `fjq.clearAll(function(err))` to clear all jobs currently in Redis. Workers currenty running won't be affected; ideally, only use this function in your test suite and never in prod!
* if you need access to a Redis connection, use `fjq.getCommandConnection()` which returns a [redis client](https://www.npmjs.com/package/redis).

## Implementation notes
* FIFO queue
* The library uses `BLPOP` to process jobs, so you'll have one Redis connection per call to `.process`
* If you never use `.create` from a worker, you won't have a second connection. From the moment you use `.create`, a new connection is established with Redis (since `BLPOP` is blocking and can't be used again) and maintained for performance.
* See https://github.com/Neamar/featureless-job-queue/issues/1 for a detailed breakdown of the performance improvement you can expect when switching from Kue to featureless-job-queue.

## Does it work?
This library is currently used in various productions environment. The biggest one I know of handles more than 45 millions tasks a day with less than 8 open Redis connections.


CHANGELOG.md
# v2.0.2
* Added `npmignore` file to make package smaller

# v2.0.0

* Removed `cargoConcurrency` option
* (breaking) `.create()` callback now returns two arguments: `err` and `count`, number of tasks in the queue.


test/index.js
"use strict";
var assert = require("assert");
var async = require("async");
var redis = require("redis");

var client = redis.createClient();
var FJQ = require('../lib');


describe("Featureless job queue", function() {
  describe("constructor", function() {
    it("should default if options is unspecified", function() {
      var fjq = new FJQ();
      assert.equal(fjq.options.redisUrl, "redis://localhost");
      assert.equal(fjq.options.redisKey, "fjq:jobs");
    });

    it("should use specified redisUrl", function() {
      var fjq = new FJQ({redisUrl: "fakeUrl"});
      assert.equal(fjq.options.redisUrl, "fakeUrl");
    });

    it("should use specified redisKey", function() {
      var fjq = new FJQ({redisKey: "fakeKey"});
      assert.equal(fjq.options.redisKey, "fakeKey");
    });
  });

  describe(".create(job, cb)", function() {
    var fjq = new FJQ();
    beforeEach(function(done) {
      fjq.clearAll(done);
    });

    it("should require a job", function(done) {
      fjq.create(null, function(err) {
        assert.ok(err.toString().indexOf("no job specified") !== -1);
        done();
      });
    });

    it("should require a valid job", function(done) {
      fjq.create(function() {}, function(err) {
        assert.ok(err.toString().indexOf("non JSON job specified") !== -1);
        done();
      });
    });

    it("should save the job to Redis", function(done) {
      var fakeJob = {foo: "bar"};
      async.waterfall([
        function createJob(cb) {
          fjq.create(fakeJob, cb);
        },
        function loadJobInRedis(count, cb) {
          client.lrange(fjq.options.redisKey, 0, 10, cb);
        },
        function checkCorrect(result, cb) {
          assert.equal(result.length, 1);
          assert.equal(result[0], JSON.stringify(fakeJob));
          cb();
        }
      ], done);
    });

    it("should save multiple jobs to Redis", function(done) {
      var fakeJob = {foo: "bar"};
      var fakeJobs = [fakeJob, fakeJob, fakeJob];

      async.waterfall([
        function createJob(cb) {
          fjq.create(fakeJobs, cb);
        },
        function loadJobInRedis(count, cb) {
          client.lrange(fjq.options.redisKey, 0, 10, cb);
        },
        function checkCorrect(result, cb) {
          assert.equal(result.length, fakeJobs.length);
          assert.equal(result[0], JSON.stringify(fakeJob));
          cb();
        }
      ], done);
    });
  });

  describe(".process(concurrency, queueWorker)", function() {
    var fjq;
    beforeEach(function(done) {
      fjq = new FJQ();
      fjq.clearAll(done);
    });

    afterEach(function(done) {
      fjq.shutdown(done);
    });

    it("should require a function", function() {
      assert.throws(function() {
        fjq.process(true);
      });
    });

    it("should work on already created tasks", function(done) {
      var fakeJob = {foo: "bar"};
      async.waterfall([
        function addJob(cb) {
          fjq.create(fakeJob, cb);
        },
        function process(count, cb) {
          fjq.process(function(job, jobCb) {
            assert.equal(job.foo, fakeJob.foo);
            jobCb();
            cb();
          }, 1);
        }
      ], done);
    });

    it("should start working on tasks queued after .process()", function(done) {
      var fakeJob = {foo: "bar"};
      var called = false;
      async.waterfall([
        function processJob(cb) {
          fjq.process(function(job, jobCb) {
            assert.equal(job.foo, fakeJob.foo);
            called = true;
            jobCb();
          }, 1);

          cb();
        },
        function addJob(cb) {
          fjq.create(fakeJob, cb);
        },
        function waitForCompletion(count, cb) {
          var interval = setInterval(function() {
            if(called) {
              clearInterval(interval);
              cb();
            }
          }, 5);
        }
      ], done);
    });

    it("should remove tasks once finished", function(done) {
      var fakeJob = {foo: "bar"};
      async.waterfall([
        function addJob(cb) {
          fjq.create(fakeJob, cb);
        },
        function processJob(count, cb) {
          fjq.process(function(job, jobCb) {
            assert.equal(job.foo, fakeJob.foo);
            jobCb();
            cb();
          }, 1);
        },
        function loadJobInRedis(cb) {
          client.lrange(fjq.options.redisKey, 0, 10, cb);
        },
        function checkCorrect(result, cb) {
          assert.equal(result.length, 0);
          cb();
        }
      ], done);
    });
  });

  describe(".length(cb)", function() {
    var fjq;
    beforeEach(function(done) {
      fjq = new FJQ();
      fjq.clearAll(done);
    });

    afterEach(function(done) {
      fjq.shutdown(done);
    });

    it("should return 0 by default", function(done) {
      fjq.length(function(err, count) {
        assert.ifError(err);
        assert.equal(count, 0);
        done();
      });
    });

    it("should return length of jobs", function(done) {
      async.waterfall([
        function createJobs(cb) {
          fjq.create([{}, {}, {}], cb);
        },
        function getLength(count, cb) {
          fjq.length(cb);
        },
        function assertLength(count, cb) {
          assert.equal(count, 3);
          cb();
        }
      ], done);

    });
  });

  describe(".shutdown(done)", function() {
    var fjq;
    beforeEach(function(done) {
      fjq = new FJQ();
      fjq.clearAll(done);
    });

    afterEach(function(done) {
      fjq.shutdown(done);
    });

    it("should prevent user from creating new tasks once shut down", function(done) {
      fjq.shutdown();
      fjq.create({}, function(err) {
        assert.ok(err.toString().indexOf("queue was shutdown") !== -1);
        done();
      });
    });

    it("should prevent user from adding new workers once shut down", function() {
      fjq.shutdown();
      assert.throws(function() {
        fjq.process(function() {}, 1);
      });
    });

    it("should only shut down once workers have finished", function(done) {
      var fakeJob = {foo: "bar"};
      var hasProcessed = false;
      async.waterfall([
        function processJob(cb) {
          fjq.process(function(job, jobCb) {
            setTimeout(function() {
              hasProcessed = true;
              jobCb();
            }, 100);
          }, 1);
          cb();
        },
        function addJob(cb) {
          fjq.create(fakeJob, cb);
        },
        function wait(count, cb) {
          setTimeout(cb, 20);
        },
        function shutdown(cb) {
          assert.equal(hasProcessed, false);
          fjq.shutdown(cb);
        },
        function ensureHasProcessed(cb) {
          assert.equal(hasProcessed, true);
          cb();
        }
      ], done);
    });
  });

  describe(".clearAll(cb)", function() {
    var fjq;
    beforeEach(function(done) {
      fjq = new FJQ();
      fjq.clearAll(done);
    });

    afterEach(function(done) {
      fjq.shutdown(done);
    });

    it("should clear all jobs", function(done) {
      async.waterfall([
        function createJobs(cb) {
          fjq.create([{}, {}, {}], cb);
        },
        function clearAll(count, cb) {
          fjq.clearAll(cb);
        },
        function getLength(removedKeys, cb) {
          assert.equal(removedKeys, 1);
          fjq.length(cb);
        },
        function assertLength(count, cb) {
          assert.equal(count, 0);
          cb();
        },

      ], done);
    });
  });
});


examples/pause.js
"use strict";
/**
 * This example demonstrate how to pause and resume a worker.
 * Please note that a paused worker will still finish all its ongoing tasks, it just won't process any more tasks
 * until you call .resume()!
 *
 * Have a look at http://caolan.github.io/async/docs.html#QueueObject for the list of things you can do on the queue.
 * Warning: do not use .shutdown() on a paused queue! shutdown() waits for all workers to finish, and would never call your callback.
 */

const JOB_COUNT = 50;
const CONCURRENCY = 1;

var FJQ = require('../lib');
var fjq = new FJQ();


// Define the worker function to use, and the concurrency
var queue = fjq.process(function workerFunction(job, cb) {
  setTimeout(function() {
    console.log("Finished job #" + job.id);

    cb();
  }, 50);
}, CONCURRENCY);


// every half second, pause or unpause the queue
// Tasks currently being worked won't be affected, but new tasks won't be processed.
setInterval(function() {
  if(queue.paused) {
    console.log("Resuming queue");
    queue.resume();
  }
  else {
    console.log("Pausing queue");
    queue.pause();
  }
}, 500).unref();


// Generate fake jobs
var jobs = [];
for(var i = 1; i <= JOB_COUNT; i += 1) {
  jobs.push({id: i});
}

// And add them to the queue
fjq.create(jobs, function(err) {
  console.log(err || "Added jobs in the queue");
});


// Only for this example: shutdown the process once all tasks have completed.
// In a real life situation you'd want to keep your worker running in case some tasks were to appear in the queue.
queue.drain = function() {
  // We can safely call .shutdown, since there is no way the queue can be drained while paused anyway.
  fjq.shutdown();
};


examples/default.js
"use strict";
/**
 * This example demonstrate the use of the library for a basic use case:
 * just enqueuing stuff in a FIFO, and working on tasks in parallel.
 * Obviously, in a real world use case, you'll probably want to run this over multiple processes or multiple servers!
 */
const JOB_COUNT = 50;
const CONCURRENCY = 5;

var FJQ = require('../lib');
var fjq = new FJQ();


// Define the worker function to use, and the concurrency
var queue = fjq.process(function workerFunction(job, cb) {
  setTimeout(function() {
    console.log("Finished job #" + job.id);

    cb();
  }, 50);
}, CONCURRENCY);


// Generate fake jobs
var jobs = [];
for(var i = 1; i <= JOB_COUNT; i += 1) {
  jobs.push({id: i});
}

// And add them to the queue
fjq.create(jobs, function(err, count) {
  console.log(err || "Added jobs in the queue, number of tasks: " + count);
});


// Only for this example: shutdown the process once all tasks have completed.
// In a real life situation you'd want to keep your worker running in case some tasks were to appear in the queue.
queue.drain = function() {
  fjq.shutdown();
};


examples/retries.js
"use strict";
/**
 * This example demonstrate how to retry a failed job.
 * The library has no built-in retry feature, since this would slow considerably most use-case,
 * but you can just requeue failed tasks.
 */
const JOB_COUNT = 50;
const CONCURRENCY = 5;

var FJQ = require('../lib');
var fjq = new FJQ();


// For this use case, assume 50% of the jobs fails randomly
function workerFunction(job, cb) {
  setTimeout(function() {
    // 50% chance of failing the job
    if(Math.random() < 0.5) {
      console.log("Finished job #" + job.id + " on attempt #" + job.attempts);
      cb();
    }
    else {
      cb(new Error("Job failed"));
    }
  }, 50);
}

// Define the worker function to use, and the concurrency
var queue = fjq.process(function workerWrapper(job, cb) {
  workerFunction(job, function(err) {
    if(err) {
      // Retry on failure.
      // You can cancel the job after a set amount of retries by not requeuing it
      job.attempts += 1;
      console.log("Requeuing job #" + job.id);
      fjq.create(job, cb);
    }
    else {
      // Successful job, just move on to next job
      cb();
    }
  });
}, CONCURRENCY);


// Generate fake jobs
var jobs = [];
for(var i = 1; i <= JOB_COUNT; i += 1) {
  jobs.push({id: i, attempts: 1});
}

// And add them to the queue
fjq.create(jobs, function(err) {
  console.log(err || "Added jobs in the queue");
});


// Only for this example: shutdown the process once all tasks have completed.
// In a real life situation you'd want to keep your worker running in case some tasks were to appear in the queue.
queue.drain = function() {
  fjq.shutdown();
};


examples/monitoring.js
"use strict";
/**
 * This example displays the use of a watchdog to monitor concurrency.
 *
 * This example also demonstrate the use of the library for a "high-concurrency" use-case:
 * 25 tasks are run in parallel, each task taking 1500ms.
 * Note that such a concurrency is only relevant for non-CPU bounds tasks,
 * e.g. network requests are a good use of such a feature since you're spending most of your time
 * waiting for the network stack anyway.
 * Monitor the CPU usage of your queue and adjust accordingly!
 * A concurrency of 200 might make sense for HTTP requests if you have a decent network adapter.
 * In any case, this library won't be an issue even if your concurrency goes over a couple millions / sec (tested and approved)
 *
 */
const CONCURRENCY = 25;

var FJQ = require('../lib');
var fjq = new FJQ();


// Define the worker function to use, and the concurrency
var queue = fjq.process(function workerFunction(job, cb) {
  setTimeout(function() {
    console.log("Finished job #" + job.id);
    cb();
  }, 1500);
}, CONCURRENCY);


// Generate fake jobs
var jobs = [];
for(var i = 0; i < 90; i += 1) {
  jobs.push({id: i});
}
jobs[jobs.length - 1].shutdown = true;


// And add them to the queue
fjq.create(jobs, function(err) {
  console.log(err || "Added jobs in the queue");
});


// Add a watcher to see how many tasks are running / in queue
setInterval(function() {
  fjq.length(function(err, length) {
    if(err) {
      throw err;
    }

    console.log("Tasks running: " + queue.running() + ", in local queue: " + queue.length() + ", in Redis queue: " + length);

    if(length === 0 && queue.workersList().length === 0) {
      process.exit();
    }
  });
}, 1000);


examples/multiple-queues.js
"use strict";
/**
 * This example demonstrate the use of the library with two different queues.
 * Each queue has a different concurrency settnig, and a different worker function.
 * In this simple example, the workers are mostly similar however the concurrency is different, and all tasks from A will finish way before B.
 * Note that the same jobs are used for both tasks, this is only for example purpose!
 * This example will use 4 connections to Redis: one for each call to .process, and one for each call to create().
 * Calling create again wouldn't add more connections, however adding more .process() would add one connection per call since the worker would keep running.
 */

const JOB_COUNT = 50;
const CONCURRENCY_TASK_A = 50;
const CONCURRENCY_TASK_B = 5;

var FJQ = require('../lib');
var fjqTaskA = new FJQ({redisKey: "taskA:jobs"});
var fjqTaskB = new FJQ({redisKey: "taskB:jobs"});


var queueA = fjqTaskA.process(function workerFunctionA(job, cb) {
  setTimeout(function() {
    console.log("Finished job from task A #" + job.id);
    cb();
  }, 50);
}, CONCURRENCY_TASK_A);


var queueB = fjqTaskB.process(function workerFunctionB(job, cb) {
  setTimeout(function() {
    console.log("Finished job from task B #" + job.id);
    cb();
  }, 50);
}, CONCURRENCY_TASK_B);


// Generate fake jobs
var jobs = [];
for(var i = 1; i <= JOB_COUNT; i += 1) {
  jobs.push({id: i});
}

// And add them to the queues
fjqTaskA.create(jobs, function(err) {
  console.log(err || "Added jobs in queue A");
});
fjqTaskB.create(jobs, function(err) {
  console.log(err || "Added jobs in queue B");
});


// Only for this example: shutdown the process once all tasks have completed.
// In a real life situation you'd want to keep your worker running in case some tasks were to appear in the queue.
queueA.drain = function() {
  fjqTaskA.shutdown();
};
queueB.drain = function() {
  fjqTaskB.shutdown();
};


examples/high-concurrency.js
"use strict";
/**
 * This example demonstrate the use of the library for a "high-concurrency" use-case:
 * 25 tasks are run in parallel, each task taking 1500ms.
 * Note that such a concurrency is only relevant for non-CPU bounds tasks,
 * e.g. network requests are a good use of such a feature since you're spending most of your time
 * waiting for the network stack anyway.
 * Monitor the CPU usage of your queue and adjust accordingly!
 * A concurrency of 200 might make sense for HTTP requests if you have a decent network adapter.
 * In any case, this library won't be an issue even if your concurrency goes over a couple millions / sec (tested and approved)
 *
 * In addition, this example also display the use of a watchdog to monitor concurrency.
 */
const CONCURRENCY = 25;

var FJQ = require('../lib');
var fjq = new FJQ();


// Define the worker function to use, and the concurrency
var queue = fjq.process(function workerFunction(job, cb) {
  setTimeout(function() {
    console.log("Finished job #" + job.id);
    cb();
  }, 1500);
}, CONCURRENCY);


// Generate fake jobs
var jobs = [];
for(var i = 0; i < 90; i += 1) {
  jobs.push({id: i});
}
jobs[jobs.length - 1].shutdown = true;


// And add them to the queue
fjq.create(jobs, function(err) {
  console.log(err || "Added jobs in the queue");
});


// Add a watcher to see how many tasks are running / in queue
setInterval(function() {
  fjq.length(function(err, length) {
    if(err) {
      throw err;
    }

    console.log("Tasks running: " + queue.workersList().length + ", in local queue: " + queue.length() + ", in Redis queue: " + length);

    if(length === 0 && queue.workersList().length === 0) {
      process.exit();
    }
  });
}, 1000);


lib/index.js
"use strict";
var async = require("async");
var redis = require("redis");
var debug = require("debug");

var log = debug("featureless-job-queue");

var noop = function() {};


/*
 * Default constructor for the library, invoke with:
 * new FJQ(options)
 * where options is an object of parameters, see README for details
 */
var FJQ = function FJQ(options) {
  options = options || {};
  options.redisUrl = options.redisUrl || "redis://localhost";
  options.redisKey = options.redisKey || "fjq:jobs";
  options.overfillRatio = options.overfillRatio || 1.1;

  this.options = options;
  this.connectionsInUse = [];
  this.queuesInUse = [];
  this.isActive = true;

  log("Initialized new featureless job queue on key " + options.redisKey);
};


/**
 * Process tasks in specified queueWorker function,
 * running up to concurrency (int) tasks at the same time
 */
FJQ.prototype.process = function process(queueWorker, concurrency) {
  if(!this.isActive) {
    throw new Error("Queue was shutdown and can't be used anymore");
  }

  if(!concurrency) {
    concurrency = 1;
  }
  if(typeof(queueWorker) !== 'function') {
    throw new Error("Queue worker must be a function!");
  }


  var self = this;

  var connection = redis.createClient(this.options.redisUrl);
  connection._fjqProcessor = true;
  this.connectionsInUse.push(connection);

  var queue = async.queue(queueWorker, concurrency);
  this.queuesInUse.push(queue);

  var refiller = function() {
    if(!self.isActive) {
      // We're shutting down this .process(), don't try to queue more tasks
      return;
    }

    connection.blpop(self.options.redisKey, 0, function(err, values) {
      if(err && err.toString().indexOf("AbortError") !== -1) {
        // Queue is being shutdown, clean everything properly
        log("Shutting down queue on " + self.options.redisKey);
        return;
      }

      if(err) {
        log(err);
        refiller();
        return;
      }
      var jobJSON = values[1];
      log("Read job: " + jobJSON);
      var job = JSON.parse(jobJSON);
      queue.push(job, refiller);
    });
  };

  // fill queue, ensuring we have a little more to be as fast as possible
  for(var i = 0; i < concurrency * this.options.overfillRatio; i += 1) {
    refiller();
  }

  return queue;
};


/**
 * Intended for internal use only, but can be used if you need to run custom redis command.
 */
FJQ.prototype.getCommandConnection = function getCommandConnection() {
  if(!this.isActive) {
    throw new Error("queue was shutdown");
  }

  // We use one connection for calls to create, and cache it.
  // If it does not already exists, create it
  if(!this.commandConnection) {
    this.commandConnection = redis.createClient(this.options.redisUrl);
    this.connectionsInUse.push(this.commandConnection);
  }

  return this.commandConnection;
};


/**
 * Add the job / jobs to the job queue, and invoke cb() when *all* tasks have been saved
 * O(n), although jobs are batched 200 at a time by default, so very likely to be O(1) for most use cases
 */
FJQ.prototype.create = function create(jobs, cb) {
  if(!this.isActive) {
    return cb(new Error("queue was shutdown"));
  }

  if(!jobs) {
    return cb(new Error("no job specified!"));
  }
  if(!Array.isArray(jobs)) {
    jobs = [jobs];
  }

  if(typeof(cb) !== 'function') {
    throw new Error("callback must be a function!");
  }

  log("Creating " + jobs.length + " job" + (jobs.length > 1 ? "s" : "") + " in " + this.options.redisKey);

  var self = this;

  var errored = false;

  var jobsJSON = jobs.map(function(job) {
    var jobJSON = JSON.stringify(job);
    if(!jobJSON) {
      errored = job;
      return "";
    }

    return jobJSON;
  });

  if(errored) {
    return cb(new Error("non JSON job specified: " + errored.toString()));
  }

  self.getCommandConnection().rpush(self.options.redisKey, jobsJSON, cb);
};


/**
 * Remove all existings jobs from this queue
 * O(n)
 */
FJQ.prototype.clearAll = function clearAll(done) {
  if(typeof(done) !== 'function') {
    throw new Error("done must be a function!");
  }
  this.getCommandConnection().del(this.options.redisKey, done);
};


/**
 * Returns the number of jobs currently stored in the job queue
 * O(1)
 */
FJQ.prototype.length = function length(done) {
  if(typeof(done) !== 'function') {
    throw new Error("done must be a function!");
  }
  this.getCommandConnection().llen(this.options.redisKey, done);
};


/**
 * Shutdown the queue, finish all the current workers jobs and exit properly.
 * Closes the Redis connection, or flush the blocked one.
 */
FJQ.prototype.shutdown = function shutdown(done) {
  var self = this;
  if(!done) {
    done = noop;
  }

  if(!this.isActive) {
    // Queue was already shut down
    return done();
  }

  this.isActive = false;

  this.connectionsInUse.forEach(function(connection) {
    if(connection._fjqProcessor) {
      // Blocking connection needs to be flushed
      connection.end(true);
    }
    else {
      // Other connection can be closed properly
      connection.quit();
    }
  });

  async.each(this.queuesInUse, function(queue, cb) {
    if(queue.idle()) {
      log("Shutting down empty queue");
      // Queue already idle
      return cb();
    }

    // Once all tasks have been handled, call cb()
    log("Shutting down currently running queue, adding listener for drain event");
    queue.drain = cb;
  }, function() {
    self.queuesInUse.forEach(function(queue) {
      queue.kill();
    });

    log("All queues shut down!");
    done();
  });
};

module.exports = FJQ;
